File: examples/typescript/test/gate.test.ts
import { describe, it, expect } from "vitest";
import { initialState, type TicketStatus } from "../src/domain";
import { fold } from "../src/fold";

function escalating(by: "Human" | "Agent" = "Agent") {
  const status: TicketStatus = { kind: "Escalating", by };
  return initialState([{ id: "4118", body: "x", priority: "High", status }]);
}

describe("the irreversible-action gate", () => {
  it("pages on a host (Human) confirm", () => {
    const [s, fx] = fold(escalating(), [{ kind: "EscalationConfirmed", ticket: "4118", by: "Human" }], 5);
    expect(s.tickets.get("4118")!.status.kind).toBe("Escalated");
    expect(fx).toEqual([{ kind: "PageOncall", ticket: "4118", at: 5 }]);
  });

  it("denies an agent self-confirm — no page fires", () => {
    const [s, fx] = fold(escalating(), [{ kind: "EscalationConfirmed", ticket: "4118", by: "Agent" }], 5);
    // status unchanged: still Escalating, never Escalated
    expect(s.tickets.get("4118")!.status.kind).toBe("Escalating");
    expect(fx).toEqual([{ kind: "Diag", note: "denied: agent self-confirm" }]);
    expect(fx.some((e) => e.kind === "PageOncall")).toBe(false);
  });
});

File: reference-impl/kotlin/src/test/kotlin/adr/GateTest.kt
package adr

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class GateTest {

    private fun escalating(by: Actor = Actor.Agent): State =
        State.initial(listOf(Ticket("4118", "x", Priority.High, TicketStatus.Escalating(by))))

    @Test
    fun `pages on a host (Human) confirm`() {
        val (s, fx) = Reducer.fold(
            escalating(),
            listOf(ToolResult.EscalationConfirmed("4118", Actor.Human)),
            5L,
        )
        assertTrue(s.tickets.getValue("4118").status is TicketStatus.Escalated)
        assertEquals(listOf(Effect.PageOncall("4118", 5L)), fx)
    }

    @Test
    fun `denies an agent self-confirm — no page fires`() {
        val (s, fx) = Reducer.fold(
            escalating(),
            listOf(ToolResult.EscalationConfirmed("4118", Actor.Agent)),
            5L,
        )
        // status unchanged: still Escalating, never Escalated
        assertTrue(s.tickets.getValue("4118").status is TicketStatus.Escalating)
        assertEquals(listOf(Effect.Diag("denied: agent self-confirm")), fx)
        assertFalse(fx.any { it is Effect.PageOncall })
    }
}
